spynnaker/pyNN/models/neuron/synaptic_manager.py

```diff
@@ -520,24 +524,32 @@ def _get_ring_buffer_to_input_left_shifts(
                     total_weights[synapse_type])
                 max_weights[synapse_type] = max(
                     max_weights[synapse_type], biggest_weight[synapse_type])
+            # This is to deal with very small weights that are floored to 0
+            mmw = 2**math.floor(math.log(min_max_weight[synapse_type], 2))
+            max_weights[synapse_type] = min(mmw * 2 ** 15,
+                                            max_weights[synapse_type])
 
         # Convert these to powers
         max_weight_powers = (
-            0 if w <= 0 else int(math.ceil(max(0, math.log(w, 2))))
+            0 if w <= 1 else int(math.ceil(max(0, math.log(w, 2))))
             for w in max_weights)
 
         # If 2^max_weight_power equals the max weight, we have to add another
         # power, as range is 0 - (just under 2^max_weight_power)!
         max_weight_powers = (
-            w + 1 if (2 ** w) <= a else w
+            w + 1 if (2 ** w) < a else w
             for w, a in zip(max_weight_powers, max_weights))
 
         # If we have synapse dynamics that uses signed weights,
         # Add another bit of shift to prevent overflows
         if weights_signed:
             max_weight_powers = (m + 1 for m in max_weight_powers)
-
-        return list(max_weight_powers)
+        rb_ls = list(max_weight_powers)
+        print("=" * 60)
+        print("RB left shifts for {:20}".format(application_vertex.label),
+              "=", rb_ls)
+        print("-" * 60)
+        return rb_ls
```

Comment:
As mentioned, this printing should be removed.  Instead it would make sense to have a report that tells the user what scaling was used and why (i.e. expected maximum weights and scaling values).

Reply:
This should also possibly warn the user up-front when weight scaling cannot be achieved due to minimum weight representation issues.  This would then explain the overflow messages that are likely to received later.

Additionally, this code could detect and warn the user when it is impossible to represent both the minimum and maximum weight, suggesting the 16-bit weight is not enough to represent the dynamic range of values requested.

Reply:
As mentioned elsewhere, we don't have the minimum weight here yet, so this might be needed if this is to report the range between the true minimum and the maximum, rather than the minimum of the maximums.